### PR TITLE
[clang] Fix CTAD not work for C++ explicit type conversion (functional annotation).

### LINKED ISSUE #64347: Aggregate CTAD of parenthesized prvalue fails
See <https://compiler-explorer.com/z/xjxWbKTs7>.
```C++
template<class T> struct X { T x; };
auto x = X(0);
```
```output
<source>:2:10: error: no viable constructor or deduction guide for deduction of template arguments of 'X'
    2 | auto x = X(0);
      |          ^
<source>:1:26: note: candidate template ignored: could not match 'X<T>' against 'int'
    1 | template<class T> struct X { T x; };
      |                          ^
<source>:1:26: note: candidate function template not viable: requires 0 arguments, but 1 was provided
1 error generated.
Compiler returned: 1
```

### PR DESCRIPTION
**Pull request by hokein**
This fixes https://github.com/llvm/llvm-project/issues/64347.

The CTAD for an aggregate class is missing to handle the explicit type conversion case, e.g. `TemplateFooClass(1, 2);`. Per C++  expr.type.conv p1, the deduced type is the return type of the deduction guide selected by the CTAD for the reminder.

In the deduction implementation `DeduceTemplateSpecializationFromInitializer`, the parenthesized express-list case relies on the `ParenListExpr` parameter (default is nullptr), the AST `ParenListExpr` node is not built for all variant initializer cases (`BuildCXXTypeConstructorExpr`, `BuildCXXNew` etc), thus the deduction doesn't perform for these cases. This patch fixes it by removing the `ParenListExpr` and using the `Inits` instead (which also simplifies the interface and implementation).

### CODE REVIEW

File: clang/lib/Sema/SemaInit.cpp [hunk 1]
```diff
@@ -10804,9 +10804,12 @@ QualType Sema::DeduceTemplateSpecializationFromInitializer(
     if (getLangOpts().CPlusPlus20 && !HasAnyDeductionGuide) {
       if (ListInit && ListInit->getNumInits()) {
         SynthesizeAggrGuide(ListInit);
-      } else if (PL && PL->getNumExprs()) {
-        InitListExpr TempListInit(getASTContext(), PL->getLParenLoc(),
-                                  PL->exprs(), PL->getRParenLoc());
+      } else if (Inits.size()) { // parenthesized expression-list
+        // Inits are expressions inside the parentheses. We don't have
+        // the parentheses source locaitons, use the begin/end of Inits as the
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaInit.cpp:10809]:
```suggestion
        // the parentheses source locations, use the begin/end of Inits as the
```

**Review by cor3ntin — COMMENTED** [review]
Why remove the support for ParenListExpr entirely?
Could you not fallback on looking at the init list expression only when ParenListExpr is null?

**Comment by hokein** [comment]
> Why remove the support for ParenListExpr entirely? Could you not fallback on looking at the init list expression only when ParenListExpr is null?

We could keep `ParenListExpr`,  but I'm not sure the benefit (IMO, removing it seems like an API improvement, and make code simpler). The `ParenListExpr` is redundant with the existing `Inits` parameter to some degree, the only benefit to use it is for the precise `()` location information (as we deduct the template type here, I think it is fine to use a less-precise source location).

Looking at all usages (total 4) of `DeduceTemplateSpecializationFromInitializer` in clang, only 1 [place](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaDecl.cpp#L12959) passes the `ParenListExpr`.

**Comment by shafik** [comment]
Can you please update your summary to explain the problem and how the fix addresses the problem.

This is usually what goes into the git log and we want those to be as descriptive as possible but also it help code reviewers to understand what they are reviewing quicker.

**Comment by shafik** [comment]
> > Why remove the support for ParenListExpr entirely? Could you not fallback on looking at the init list expression only when ParenListExpr is null?
> 
> We could keep `ParenListExpr`, but I'm not sure the benefit (IMO, removing it seems like an API improvement, and make code simpler). The `ParenListExpr` is redundant with the existing `Inits` parameter to some degree, the only benefit to use it is for the precise `()` location information (as we deduct the template type here, I think it is fine to use a less-precise source location).
> 
> Looking at all usages (total 4) of `DeduceTemplateSpecializationFromInitializer` in clang, only 1 [place](https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaDecl.cpp#L12959) passes the `ParenListExpr`.

As long as it does not effect the quality of the diagnostics then this makes sense.

**Review by shafik — APPROVED** [review]
LGTM

**Comment by hokein** [comment]
> Can you please update your summary to explain the problem and how the fix addresses the problem.
> 
> This is usually what goes into the git log and we want those to be as descriptive as possible but also it help code reviewers to understand what they are reviewing quicker.

Thanks, I have updated the description with more details.

**Comment by JohelEGP** [comment]
> Thanks, I have updated the description with more details.

If it has to appear in the git log, it has to be added to the commit message.
GitHub is separate from Git, so updating the opening comment of the PR won't be visible through `git`.

**Comment by hokein** [comment]
> > Thanks, I have updated the description with more details.
> 
> If it has to appear in the git log, it has to be added to the commit message. GitHub is separate from Git, so updating the opening comment of the PR won't be visible through `git`.

GitHub's "squash and merge" button already does this for us.

**Pull request merged**